### PR TITLE
Add the Count-Min Sketch and introduce the new sketch module

### PR DESCRIPTION
**Pull request by smastelini**
Here is the [roadmap page](https://maxhalford.notion.site/Implement-Min-Count-Sketch-b85ee293b7f44c9982dbceff3341715f) 😃 .

CMS is a classic one, and I'm happy to finally bring it aboard!

**Edit:** the PR ended up becoming bigger.

We are showing some love for really useful summarization and data sketching structures. All of them now lie in the ~`collections`~ `sketch` module. Thanks, @MaxHalford, for the naming suggestions!

### CODE REVIEW

**Comment by MaxHalford** [comment]
> I have also added the dot product capabilities described in the paper, but I'm not sure how to explore that.

A cool thing you can do: count the number of times a user has seen a move, and have one CMS per user to do that to save space. Then, calculate the dot-product between two user CMS to estimate their cosine similarity. It's kind of the basis of factorization machines. @gbolmier what do you think?

**Comment by smastelini** [comment]
Yeah! I makes a lot of sense :)

**Comment by gbolmier** [comment]
It sounds interesting but I'm not sure to see the benefits. Techniques that learn latent factors like Factorization Machines already compress users representations very well and in a meaningful way. Building CMS representations would be faster than learning latent vectors but it doesn't matter when you learn in an async manner. I didn’t know about CMS five minutes ago so please lmk if I'm completely missing the point.

File: river/stats/cms.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        return sum(self._cms[0])
+
+    def __getitem__(self, x) -> int:
+        # Point query
+        return min(self._cms[self._hash(x)])
+
+    def __matmul__(self, other) -> int:
+        # Dot product
+        return min(np.einsum("ij,ij->i", self._cms, other._cms))
+
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:163]: I thought about that, but `x` could be any hashable. So I'm not sure how to deal with random input types where ordering is not explicitly. 

Do you have any tips?

File: river/stats/cms.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    def _hash(self, x):
+        return tuple(zip(*((i, (hash(x) ^ self._masks[i]) % self._w) for i in range(self._d))))
+
+    def update(self, x: typing.Hashable, w: int = 1):
+        self._cms[self._hash(x)] += w
+
+        return self
+
+    def get(self):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:152]: I agree with your point. I thought about passing a parameter to `get` but then we break the standards followed so far. Besides that, [] is fancier and straightforward.

File: river/stats/cms.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxHalford) [reply river/stats/cms.py:163]: You have a good point! Would the `cms[a:b]` operation provide a speed-up over `sum(cms[x] for x in range(a, b))`? If not then I simply wouldn't add it. If it does, then maybe we can just try handling integers and document this clearly.

File: river/stats/cms.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        self.epsilon = epsilon
+        self.delta = delta
+        self.seed = seed
+
+        self._w = int(math.ceil(math.e / self.epsilon))
+        self._d = int(math.ceil(math.log(1 / self.delta)))
+
+        self._rng = random.Random(self.seed)
+        self._masks = [self._rng.getrandbits(64) for _ in range(self._d)]
+        self._cms = np.zeros((self._d, self._w), dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:142]: I assumed vectorization would be nice here. I can check if lists are better.

File: river/stats/cms.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    >>> from river import stats
+
+    >>> cms = stats.CMS(epsilon=0.005, seed=0)
+
+    >>> # To generate random numbers
+    >>> rng = random.Random(7)
+    >>> counter = collections.Counter()
+
+    We can check the number of slots per hash table:
+    >>> cms.w
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:59]: n_tables or n_hash could also be a good choice to replace `d`

File: river/stats/cms.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:142]: If we really want to save memory, `collections.defaultdict(int)` might be a possibility too. What do you think, @MaxHalford ?

File: river/stats/cms.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    >>> cms[7]
+    12
+
+    >>> counter[532]
+    4
+
+    >>> cms[532]
+    15
+
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:86]: Sure! I'll do that asap

File: river/stats/cms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:59]: On the other hand, now I realize that `w` and `d` are well documented. I even put the equations to obtain these values. In this case I still vote to match the original paper. All the descriptions of the algorithm I saw use the same notation, so I think it is fine.

File: river/stats/cms.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
+import math
+import random
+import typing
+
+import numpy as np
+
+from river import stats
+
+
+class CMS(stats.base.Univariate):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py:10]: You have a good point. I'll rename it :)

File: river/stats/cms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaxHalford) [reply river/stats/cms.py:59]: No problem

File: river/stats/cms.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        """Return the total sum of monitored counts."""
+        return sum(self._cms[0])
+
+    def __getitem__(self, x) -> int:
+        # Point query
+        return min(self._cms[self._hash(x)])
+
+    def __matmul__(self, other) -> int:
+        # Dot product
+        return min(np.einsum("ij,ij->i", self._cms, other._cms))
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py]: Done!!

**Comment by smastelini** [comment]
@MaxHalford and I agree that Count-Min sketch and Lossy Count are not exactly statistics. Not in the same sense as those in `river.stats`. For that reason, for the time being, they will be moved to `river.misc`. 

This gives us more freedom to provide a more concise API for these sketch algorithms.

TODO:

- [x] Move `CountMin` to `misc`
- [ ] Move `LossyCount` to `misc`
- [ ] Update release notes

File: river/stats/cms.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py]: Moving `CountMin` to another module will do the trick. We will provide a more concise API for this kind of tool. The same goes for `LossyCount` :D

File: river/stats/cms.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py]: I went for `CountMin` but we can add the `Sketch` suffix if needed.

File: river/stats/cms.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    544
+
+    And the number of hash tables:
+    >>> cms.d
+    3
+
+    Let's compare the sketch against a brute force approach:
+
+    >>> for _ in range(10000):
+    ...     v = rng.randint(-1000, 1000)
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/stats/cms.py]: I am out of good ideas for that. Do you have any suggestions?

File: river/misc/cms.py [hunk 9]
```diff
@@ -119,7 +125,7 @@ class CountMin:
     And use it to calculate the cosine distance between the elements monitored in `cms_a` and `cms_b`
 
     >>> cosine_dist(cms_a, cms_b)
-    0.8785023776056446
+    0.17536377918173912
```
Comment (MaxHalford) [inline river/misc/cms.py:132]:
```suggestion
    0.175363...
```

It leaves more room for numerical errors with different Python/dependency versions

File: river/misc/cms.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        if self.n_slots != other.n_slots or self.n_tables != other.n_tables:
+            unmatched_dims = (
+                f"({self.n_slots}, {self.n_tables}) vs. ({other.n_slots}, {other.n_tables})"
+            )
+            raise ValueError(f"The number of slots and hash tables do not match: {unmatched_dims}.")
+
+        return min(np.einsum("ij,ij->i", self._cms, other._cms))
+
+    def __len__(self):
+        return self.n_slots * self.n_tables
```
Comment (MaxHalford) [inline river/misc/cms.py:172]: Can you add a comment to indicate what this corresponds to?

File: river/misc/cms.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+    def update(self, x: typing.Hashable, w: int = 1):
+        self._cms[self._hash(x)] += w
+
+    def total(self) -> int:
+        """Computes the total sum of keys."""
```
Comment (MaxHalford) [inline river/misc/cms.py:178]:
```suggestion
        """Return the total count."""
```

File: river/misc/cms.py (hunk 10 above)
Comment (smastelini) [reply river/misc/cms.py]: Done!

File: river/misc/cms.py (hunk 9 above)
Comment (smastelini) [reply river/misc/cms.py:134]: I didn't know this trick!!! Nice :D

File: river/collections/counter.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    ...     cms_b = cms_b.update(v)
+
+    Now, we can define a cosine distance function:
+
+    >>> def cosine_dist(cms_a, cms_b):
+    ...     num = cms_a @ cms_b
+    ...     den = math.sqrt(cms_a @ cms_a) * math.sqrt(cms_b @ cms_b)
+    ...     return num / den
+
+    And use it to calculate the cosine distance between the elements monitored in `cms_a` and `cms_b`
```
Comment (smastelini) [inline river/collections/counter.py:132]:
```suggestion
    And use it to calculate the cosine distance between the elements monitored in `cms_a` and `cms_b`:
```

File: docs/releases/unreleased.md [hunk 13]
```diff
@@ -1,7 +1,10 @@
...
-- Move `stats.LossyCount` to `misc.LossyCount` and update its API to better match `collections.Counter`.
-- Added missing return `self` in `LossyCount`.
-- Added the Count-Min Sketch (`CountMin`) algorithm for approximate element counting.
\ No newline at end of file
+## collections
+
+- Rename `river.misc` to `river.collections`.
+- Move `stats.LossyCount` to `river.collections.HeavyHitters` and update its API to better match `collections.Counter` from the standard Python library.
+- Added missing return `self` in `HeavyHitters`.
+- Added the Count-Min Sketch (`river.collections.Counter`) algorithm for approximate element counting.
```
Comment (MaxHalford) [inline docs/releases/unreleased.md:10]: After having slept on it, I think `LossyCounter` would be a good name. My issue with `Counter` is that some users will think it's the same as that of the standard library.

File: river/collections/counter.py [hunk 14]
```diff
@@ -47,9 +47,9 @@ class CountMin(base.Base):
     Examples
     --------
     >>> import collections
-    >>> from river import misc
+    >>> import river.collections as rvc
```
Comment (MaxHalford) [inline river/collections/counter.py:50]:
```suggestion
    >>> from river import collections as rvc
```

**Review by MaxHalford — APPROVED** [review]
Looking good! Approved with a renaming suggestion

File: docs/releases/unreleased.md (hunk 13 above)
Comment (smastelini) [reply docs/releases/unreleased.md:10]: I think we have a small problem:

- Lossy Count is a heavy hitters trackers (currently `HeavyHitters`)
- Count-min is for counting (currently `Counter`)

If we rename Counter to LossyCounter, users might think that Count-min is actually Lossy Count 😅

To be fair, I'm still not totally happy with `collections`. I think it is indeed misleading, although straightforward. 

`sketch` or `containers` (or another thing) could give the idea of approximation, summarization. 

Any suggestions?

Comment (smastelini) [reply docs/releases/unreleased.md:10]: `summary`?

Comment (MaxHalford) [reply docs/releases/unreleased.md:10]: Ok maybe `sketch` is a good idea then. We can say it contains lossy containers/collections

Comment (MaxHalford) [reply docs/releases/unreleased.md:10]: And then yeah we call CountMin `Counter`

**Pull request merged**